lib/jsonrpc_client.ml

```diff
@@ -17,49 +17,102 @@
+let timeout_read fd timeout =
+	let buf = Buffer.create !json_rpc_max_len in
+	let read_start = Mtime_clock.counter () in
+	let get_total_used_time () = Mtime.Span.to_uint64_ns (Mtime_clock.count read_start) in
+	let rec inner max_time max_bytes =
+		let (ready_to_read, _, _) = try Unix.select [fd] [] [] (to_s max_time) with
+			(* in case the unix.select call fails in situation like interrupt *)
+			| _ -> [], [], []
+		in
+		(* This is not accurate the calculate time just for the select part. However, we
+		 * think the read time will be minor comparing to the scale of tens of seconds.
+		 * the current style will be much concise in code. *)
+		let remain_time = 
+			let used_time = get_total_used_time () in
+			Int64.sub timeout used_time
+		in
+		if remain_time < 0L then 
 		begin
-			match c with
-			| '{' when not !in_string -> brace_cnt := !brace_cnt + 1
-			| '}' when not !in_string -> brace_cnt := !brace_cnt - 1
-			| '"' when !in_string && (last_char () <> '\\') -> in_string := false
-			| '"' when not !in_string -> in_string := true
-			| _ -> ()
+			debug "Timeout after read %d" (Buffer.length buf);
+			raise Timeout
 		end;
-		Buffer.add_char buf c;
-		if !brace_cnt > 0 then
-			get ()
+		if List.mem fd ready_to_read then
+		begin
+			let bytes = Bytes.make 4096 '\000' in
+			match Unix.read fd bytes 0 4096 with
+				| 0 -> Buffer.contents buf (* EOF *)
+				| n ->
+					if n > max_bytes then
+					begin
+						debug "exceeding maximum read limit %d, clear buffer" !json_rpc_max_len;
+						Buffer.clear buf;
+						raise Read_error
+					end
+					else
+					begin
+						Buffer.add_subbytes buf bytes 0 n;
+						inner remain_time (max_bytes - n)
+					end
+				| exception Unix.Unix_error((Unix.EAGAIN | Unix.EWOULDBLOCK | Unix.EINTR),_,_) ->
+					inner remain_time max_bytes
+		end
+		else inner remain_time max_bytes
 	in
-	get ();
-	Buffer.contents buf
-
-let receive fin =
-	let obj = input_json_object fin in
-	debug "Response: %s" obj;
-	Jsonrpc.response_of_string obj
+	inner timeout !json_rpc_max_len
 
-let with_connection sockaddr f =
-	let fin, fout = Unix.open_connection sockaddr in
-	debug "Connected.";
-	let result = f fin fout in
-	Unix.shutdown_connection fin;
-	close_in fin;
-	debug "Shut down.";
-	result
+(* Write as many bytes to a file descriptor as possible from data before a given clock time. *)
+(* Raises Timeout exception if the number of bytes written is less than the specified length. *)
+(* Writes into the file descriptor at the current cursor position. *)
+let timeout_write filedesc total_length data response_time =
+	let write_start = Mtime_clock.counter () in
+	let get_total_used_time () = Mtime.Span.to_uint64_ns (Mtime_clock.count write_start) in
+	let rec inner_write offset max_time =
+		let (_, ready_to_write, _) = try Unix.select [] [filedesc] [] (to_s max_time) with
+			(* in case the unix.select call fails in situation like interrupt *)
+			| _ -> [], [], []
```

Comment:
I'm worried about the `_` here and think we should just catch `EINTR`, because that is an error that it is worth immediately retrying on. If an other, more persistent, error occurs, then we'll end up spinning in this loop until the timeout occurs. It's better to let the exception break the loop in that case.